### DISCUSSION TITLE
Define Queryable Proxy Model Independent of Managed Database

### DISCUSSION THREAD

**Discussion by pykenny** [sqlalchemy/sqlalchemy#8757] [Usage Questions]
Is it possible to create a model that's independent of the database managed by SA engine?

Background: Our application uses Google BigQuery that occasionally issue queries integrating with external database through [external connection and `EXTERNAL_QUERY()` function](https://cloud.google.com/bigquery/docs/reference/standard-sql/federated_query_functions#external_query). (BigQuery will handle type conversion in between, so output schema should be predictable)

For instance, the below query tries to find out records at BigQuery's end that haven't been synced with external resource recently:

```sql
SELECT I.id
FROM `my_database.internal` AS I
JOIN (
  SELECT
    id, updated_at
  FROM
    EXTERNAL_QUERY("my_external_source", "SELECT id, updated_at FROM some_table")
) AS E ON I.id = E.id
WHERE I.updated_at < E.updated_at
ORDER BY I.id
```

_2022/11/05 Update_: Some more details about the example above -- in this query, `my_database.internal` points to a managed BigQuery table that can simply be modeled through regular SA usage. In contrast, `some_table` is a table lies in external/remote database, connected and queried through `EXTERNAL_QUERY()` function. The external database usually has different system / query dialect and may not be practical to build another set of tables/models reflecting the whole database.

It would be nice if we can wrap the subquery into a model that exposes `id` and `updated_at` fields, so that it can be queried through ORM (declarative) interface and combined with other managed models. When get compiled, it can generate the `EXTERNAL_QUERY()` statement through hard-coding or some SQL query generator internally. Could be better if it can also be parameterized so that `some_table` in the query can be swapped with other table names like `some_other_table` when we want to repeat the same process on other tables.

Designing a customized function (and extend it into corresponding subquery when compiled) may be another solution, but not sure if it's possible to design a function that returns a subquery-like result with specified schema.

**Comment by zzzeek** [comment]
> Is it possible to create a model that's independent of the database managed by SA engine?
> 
> Background: Our application uses Google BigQuery that occasionally issue queries integrating with external database through [external connection and `EXTERNAL_QUERY()` function](https://cloud.google.com/bigquery/docs/reference/standard-sql/federated_query_functions#external_query). (BigQuery will handle type conversion in between, so output schema should be predictable)
> 
> For instance, the below query tries to find out records at BigQuery's end that haven't been synced with external resource recently:
> 
> ```sql
> SELECT I.id
> FROM `my_database.internal` AS I
> JOIN (
>   SELECT
>     id, updated_at
>   FROM
>     EXTERNAL_QUERY("my_external_source", "SELECT id, updated_at FROM some_table")
> ) AS E ON I.id = E.id
> WHERE I.updated_at < E.updated_at
> ORDER BY I.id
> ```
> 
> In the above example, it would be nice if we can wrap the subquery into a model that exposes `id` and `updated_at` fields, so that it can be queried through ORM (declarative) interface and combined with other managed models. When get compiled, it can generate the `EXTERNAL_QUERY()` statement through hard-coding or some SQL query generator internally. Could be better if it can also be parameterized so that `some_table` in the query can be swapped with other table names like `some_other_table` when we want to repeat the same process on other tables.

that's fine, just create a model and/or set of Table objects that expresses `some_table` and related structures that are in this remote system.    As for the "EXTERNAL_QUERY" thing, I think that could be integrated using [@compiles](https://docs.sqlalchemy.org/en/14/core/compiler.html).  I guess in this case it's a custom `FromClause` and those are generally non-obvious so see the POC below.

```py
from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.orm import declarative_base
from sqlalchemy.sql.selectable import Subquery


class EXTERNAL_QUERY(Subquery):
    def __init__(self, database, *arg, **kw):
        self.database = database

        # this is a little bit of internals, there's other ways to
        # build this object but this was most expedient
        self._init(*arg, **kw)


@compiles(EXTERNAL_QUERY)
def _render_external_query(element, compiler, **kw):
    return f'EXTERNAL_QUERY({element.database}, "{compiler.visit_select(element.element, **kw)}")'


Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)


entity = EXTERNAL_QUERY(
    "some database", select(A.id, A.data).where(A.data == "foo")
)
stmt = select(entity)

print(stmt)

# it's a Subquery, so it has a ".c" collection.  you can map it to an ORM
# class also 
stmt = select(entity.c.id, entity.c.data).where(entity.c.data == "foo")

print(stmt)

```

**Reply by pykenny** [reply]
Do we need to create an empty table for model `A` in the managed database before schema migration? I guess without doing that, Alembic will suggests creating table `a` anytime we auto-generate migration script. (Since Alembic thinks `A` should be mapped to a table named "`a`")

**Reply by pykenny** [reply]
Oh wait, but we can [create filter](https://alembic.sqlalchemy.org/en/latest/autogenerate.html#controlling-what-to-be-autogenerated) on which parts should be investigated during auto-generation phase, is that right?

**Reply by CaselIT** [reply]
Yes, alembic has multiple filters that you can use to control what it considers when running autogenerate

**Reply by pykenny** [reply]
Another question: what if the `EXTERNAL_QUERY()` part uses different dialect (I think in most cases MySQL or PostgreSQL) in query parameter ? Is it possible to use another compiler (with dialect of the queried database) when processing the component?

**Reply by zzzeek** [reply]
> Another question: what if the `EXTERNAL_QUERY()` part uses different dialect (I think in most cases MySQL or PostgreSQL) in query parameter ? Is it possible to use another compiler (with dialect of the queried database) when processing the component?

you could do that, sure, a little awkward but call `.compile()` on the inner statement directly to get its string, passing a dialect object, see https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#stringifying-for-specific-databases
